pkg/openapiconv/convert.go

```diff
@@ -0,0 +1,270 @@
+		return nil
+	}
+	return &spec3.ExternalDocumentation{
+		ExternalDocumentationProps: spec3.ExternalDocumentationProps{
+			Description: v2ED.Description,
+			URL:         v2ED.URL,
+		},
+	}
+}
+
+func ConvertComponents(v2SecurityDefinitions spec.SecurityDefinitions, v2Definitions spec.Definitions, v2Responses map[string]spec.Response, produces []string) *spec3.Components {
+	components := &spec3.Components{
+		Schemas:         make(map[string]*spec.Schema),
+		SecuritySchemes: make(spec3.SecuritySchemes),
+		Responses:       make(map[string]*spec3.Response),
+	}
+	for s, schema := range v2Definitions {
+		components.Schemas[s] = ConvertSchema(&schema)
+	}
+
+	for s, securityScheme := range v2SecurityDefinitions {
+		components.SecuritySchemes[s] = ConvertSecurityScheme(securityScheme)
+	}
+	for r, response := range v2Responses {
+		components.Responses[r] = ConvertResponse(&response, produces)
+	}
+
+	return components
+}
+
+func ConvertSchema(v2Schema *spec.Schema) *spec.Schema {
+	if v2Schema == nil {
+		return nil
+	}
+	var v3Schema spec.Schema
+	v3Schema = *v2Schema
+
+	if refString := v2Schema.Ref.String(); refString != "" {
+		if idx := strings.Index(refString, OpenAPIV2DefPrefix); idx != -1 {
+			v3Schema.Ref = spec.MustCreateRef(OpenAPIV3DefPrefix + refString[idx+len(OpenAPIV2DefPrefix):])
+		} else {
+			klog.Errorf("Error: Swagger V2 Ref %s does not contain #/definitions\n", refString)
+		}
+	}
+
+	for key, property := range v2Schema.Properties {
+		v3Schema.Properties[key] = *ConvertSchema(&property)
+	}
+	if v2Schema.Items != nil {
+		v3Schema.Items.Schema = ConvertSchema(v2Schema.Items.Schema)
+		if len(v2Schema.Items.Schemas) > 0 {
+			v3Schema.Items.Schemas = []spec.Schema{}
+			for _, s := range v2Schema.Items.Schemas {
+				v3Schema.Items.Schemas = append(v3Schema.Items.Schemas, *ConvertSchema(&s))
+			}
+		}
+	}
+
+	if v2Schema.AdditionalProperties != nil {
+		v3Schema.AdditionalProperties.Schema = ConvertSchema(v2Schema.AdditionalProperties.Schema)
+	}
+	if v2Schema.AdditionalItems != nil {
+		v3Schema.AdditionalProperties.Schema = ConvertSchema(v2Schema.AdditionalProperties.Schema)
+	}
+
+	return &v3Schema
+}
+
+func ConvertSecurityScheme(v2securityScheme *spec.SecurityScheme) *spec3.SecurityScheme {
+	if v2securityScheme == nil {
+		return nil
+	}
+	securityScheme := &spec3.SecurityScheme{
+		VendorExtensible: v2securityScheme.VendorExtensible,
+		SecuritySchemeProps: spec3.SecuritySchemeProps{
+			Description: v2securityScheme.Description,
+			Type:        v2securityScheme.Type,
+			Name:        v2securityScheme.Name,
+			In:          v2securityScheme.In,
+		},
+	}
+
+	if v2securityScheme.Flow != "" {
+		securityScheme.Flows = make(map[string]*spec3.OAuthFlow)
+		securityScheme.Flows[v2securityScheme.Flow] = &spec3.OAuthFlow{
+			OAuthFlowProps: spec3.OAuthFlowProps{
+				AuthorizationUrl: v2securityScheme.AuthorizationURL,
+				TokenUrl:         v2securityScheme.TokenURL,
+				Scopes:           v2securityScheme.Scopes,
```

Comment:
This shares a reference between the two instances on `Scopes`. Is this conversion intended to create a copy? If not, I think it should be documented that references will be shared. There are likely other places  in this conversion where references are being shared.

Reply:
Added documentation. I know we probably shouldn't be sharing the references but this function is intended to be used only by the aggregator via json Unmarshal -> v2 -> v3 -> json Marshal so it may be fine to keep as is. Let me know if there's a preference